Example of GraphQL, fixes #306

rdebusscher opened the pull request.

Karm commented:
/me looking...

Karm reviewed (approved):
@rdebusscher thank you so much :)   
What I did so far is that I built and run starter and tried to generate a project. Right off the bat, I had trouble understanding that I had to select Liberty to enable the option (it was grayed out, disabled). When I looked into the code, I switched to Liberty and downloaded it.

Could we somehow communicate to users why s the example disabled when they open the site?  (we assume we are dealing with people who are not knowledgeable about our particular runtimes )

Index.html is correct, but there is no paragraph about the example in the generated readme file. 

I tried the example app copy-paste with Quarkus 1.7.5 and it works just fine.   So I will follow up on that I guess with next PR.

:+1:

Karm reviewed (changes requested):
pls, fix checkstyle warnings...

rdebusscher commented:
Updated the PR with

- checkbox only is displayed when runtime supports it
- SelectAll / ClearAll works also for the standalone specs
- Added a section in the generated readme.md for GraphQL.
- Fixed checkstyle issues

Emily-Jiang commented on src/main/java/org/eclipse/microprofile/starter/addon/microprofile/servers/model/StandaloneMPSpec.java:
```diff
@@ -0,0 +1,106 @@
+    // @formatter:off
+    GRAPH_QL("graphql", "Graph QL",
+            "https://github.com/eclipse/microprofile-graphql/releases/tag/%s",
+            "Graph QL: GraphQL is a query language for APIs and a runtime for fulfilling those queries with your existing data.",
```
a runtime =>  a library

andymc12 replied:
Would it be possible to change the textual references to "GraphQL" (without the space)?  In the code, the `GRAPH_QL` with the underscore is fine, but in the user-facing text, it should be "GraphQL" to match both the specification from Facebook and the Microprofile project name.

Emily-Jiang commented on src/main/resources/files/DuplicateSuperHeroException.java.tpl:
```diff
@@ -0,0 +1,9 @@
+package [# th:text="${java_package}"/].graphql.db;
```
Can you add copyright statement?

rdebusscher replied:
Template files and generated code files do no need copyright statements (the other .tpl files doesn't have them either)

Emily-Jiang commented on src/main/resources/files/HeroDatabase.java.tpl:
```diff
@@ -0,0 +1,164 @@
+package [# th:text="${java_package}"/].graphql.db;
```
Can you add copyright statement?

Emily-Jiang commented on src/main/resources/files/HeroFinder.java.tpl:
```diff
@@ -0,0 +1,155 @@
+package [# th:text="${java_package}"/].graphql;
```
Can you add copyright statement?

Emily-Jiang commented on src/main/resources/files/SuperHero.java.tpl:
```diff
@@ -0,0 +1,71 @@
+package [# th:text="${java_package}"/].graphql.model;
```
Can you add copyright statement?

Emily-Jiang commented on src/main/resources/files/Team.java.tpl:
```diff
@@ -0,0 +1,72 @@
+package [# th:text="${java_package}"/].graphql.model;
```
Can you add copyright statement?

Emily-Jiang commented on src/main/resources/files/UnknownCharacterException.java.tpl:
```diff
@@ -0,0 +1,10 @@
+package [# th:text="${java_package}"/].graphql.model;
```
Can you add copyright statement?

Emily-Jiang commented on src/main/resources/files/UnknownHeroException.java.tpl:
```diff
@@ -0,0 +1,9 @@
+package [# th:text="${java_package}"/].graphql.db;
```
Can you add copyright statement?

Emily-Jiang commented on src/main/resources/files/UnknownTeamException.java.tpl:
```diff
@@ -0,0 +1,9 @@
+package [# th:text="${java_package}"/].graphql.db;
```
Can you add copyright statement?

Emily-Jiang commented on src/main/resources/files/mp3_x/index.html.tpl:
```diff
@@ -53,5 +53,10 @@ If you have a tracer running, you can open it now.
 [/]
 
+[# th:if="${mp_graphql}"]
+<h3>Rest Client</h3>
```
should be <h3>Graph QL</h3>

andymc12 replied:
```suggestion
<h3>GraphQL</h3>
```

Emily-Jiang commented on src/main/resources/files/readme.md.tpl:
```diff
@@ -107,4 +107,12 @@ Open [http://localhost:16686/](http://localhost:16686/) to see the traces. You h
+[# th:if="${mp_graphql}"]
+### GraphQL
+
+GraphQL is a query language for APIs and a runtime for fulfilling those queries with your existing data. Specification [here](https://microprofile.io/project/eclipse/microprofile-graphql)
```
a runtime => a library

Emily-Jiang commented on src/main/resources/files/readme.md.tpl:
```diff
@@ -107,4 +107,12 @@ Open [http://localhost:16686/](http://localhost:16686/) to see the traces. You h
+
+GraphQL is a query language for APIs and a runtime for fulfilling those queries with your existing data. Specification [here](https://microprofile.io/project/eclipse/microprofile-graphql)
+
+The example contains a limited version of the SuperHero example from the TCK and shows the Schema file for it.
```
I think we need to put more info to describe how the sample works. @andymc12 can you provide some texts to go into the readme so that anyone can follow the instruction and experiment with graphQL to put in queries and see the output?

andymc12 replied:
@Emily-Jiang I agree - would it make sense for me to open a separate issue and provide the text there (or in a separate PR)? There is a lot of good work in this PR, and I wouldn't want to hold it up on me writing a good example for the readme.

Emily-Jiang replied:
sure, go ahead to open a new PR

Emily-Jiang commented:
Another thought: do we need to update the file index.html to list some info about GraphQL?

andymc12 commented on src/main/java/org/eclipse/microprofile/starter/addon/microprofile/servers/MicroprofileServersAddon.java:
```diff
@@ -117,6 +142,10 @@ private String invalidSpecValue(List<String> invalidSpecs) {
     }
 
+    private String invalidStandaloneSpecValue(List<String> invalidSpecs) {
+        return "Unknown value for option 'mp.stabdaloneSpecs' : " + String.join(", ", invalidSpecs);
```
```suggestion
        return "Unknown value for option 'mp.standaloneSpecs' : " + String.join(", ", invalidSpecs);
```
typo

andymc12 commented on src/main/resources/files/liberty/server.xml.tpl:
```diff
@@ -3,6 +3,9 @@
+        <feature>mpGraphQL-1.0</feature>
+        [/]
     </featureManager>
 
```
Could we also add this line:
```
<variable name="io.openliberty.enableGraphQLUI" value="true" />
```

This will enable the GraphiQL user interface which will make it easier for users to quickly write and execute queries/mutations.  It should be inserted at the same XML level as `<featureManager>` and `<httpEndpoint>`.

andymc12 reviewed (commented):
This is awesome!  Thanks for putting this together!  I have made a few comments/suggestions, but otherwise, this looks good to me!

rdebusscher commented:
New commit with review comments addressed

andymc12 reviewed (approved):
LGTM - thanks @rdebusscher!
